demos/multigrid/geometric_multigrid.py.rst

```diff
@@ -0,0 +1,248 @@
+
+  x, y = SpatialCoordinate(mesh)
+
+  f = -0.5*pi*pi*(4*cos(pi*x) - 5*cos(pi*x*0.5) + 2)*sin(pi*y)
+
+  L = f*v*dx
+
+The exact solution is::
+
+  exact = sin(pi*x)*tan(pi*x*0.25)*sin(pi*y)
+
+We'll demonstrate a few different sets of solver parameters, so let's define a
+function that takes in set of parameters and returns the solution ::
+
+  def run_solve(parameters):
+      u = Function(V)
+      solve(a == L, u, bcs=bcs, solver_parameters=parameters)
+      return u
+
+and another to compute the error. ::
+
+  def error(u):
+      expect = Function(V).interpolate(exact)
+      return norm(assemble(u - expect))
+
+Specifying the solver
+~~~~~~~~~~~~~~~~~~~~~
+
+Let's start with our first test.  We'll confirm a working solve by
+using a direct method. ::
+
+  u = run_solve({"ksp_type": "preonly", "pc_type": "lu"})
+  print 'LU solve error', error(u)
+
+Next we'll use the conjugate gradient method preconditioned by a
+geometric multigrid V-cycle.  Firedrake automatically takes care of
+rediscretising the operator on coarse grids, and providing the number
+of levels to PETSc. ::
+
+  u = run_solve({"ksp_type": "cg", "pc_type": "mg"})
+  print 'MG V-cycle + CG error', error(u)
+
+For such a simple problem, an appropriately configured multigrid solve
+can achieve algebraic error equal to discretisation error in one
+cycle, without the application of a Krylov accelerator.  In
+particular, for the Poisson equation with constant coefficients, a
+single full multigrid cycle with appropriate chosen smoothers achieves
+discretisation error.  As ever, PETSc allows us to configure the
+appropriate settings using solver parameters. ::
+
+  parameters = {
+     "ksp_type": "preonly",
+     "pc_type": "mg",
+     "pc_mg_type": "full",
+     "mg_levels_ksp_type": "chebyshev",
+     "mg_levels_ksp_max_it": 2,
+     "mg_levels_pc_type": "jacobi"
+  }
+
+  u = run_solve(parameters)
+  print 'MG F-cycle error', error(u)
+     
+A saddle-point system: The Stokes equations
+-------------------------------------------
+
+Having demonstrated basic usage, we'll now move on to an example where
+the configuration of the multigrid solver is somewhat more complex.
+This demonstrates how the multigrid functionality composes with the
+other aspects of solver configuration, like fieldsplit
+preconditioning.  We'll use Taylor-Hood elements and solve a problem
+with specified velocity inflow and outflow conditions. ::
+
+  mesh = RectangleMesh(15, 10, 1.5, 1)
+
+  hierarchy = MeshHierarchy(mesh, 3)
+
+  mesh = hierarchy[-1]
+
+  V = VectorFunctionSpace(mesh, "CG", 2)
+  W = FunctionSpace(mesh, "CG", 1)
+  Z = V * W
+
+  u, p = TrialFunctions(Z)
+  v, q = TestFunctions(Z)
+
+  a = (inner(grad(u), grad(v)) - p * div(v) + div(u) * q)*dx
+
+  L = inner(Constant((0, 0)), v) * dx
+
+  x, y = SpatialCoordinate(mesh)
+
+  t = conditional(y < 0.5, y - Constant(1.0/4.0), y - Constant(3.0/4.0))
+  l = Constant(1.0/6.0)
+  gbar = conditional(Or(And(Constant(1.0/4.0) - l/2 < y,
```

Comment:
I don't quite see the point of wrapping all these numbers with `Constant` in this example, especially since they are inlined in the expression so you won't be able to change their values later. (Reminder: these are functions constant in space values which are passed in to assembly kernels rather than constant values hardcoded in the kernel.)